ouroboros-consensus/docs/VersioningSchemeDecision.md

```diff
@@ -0,0 +1,286 @@
+
+PRs do not alter the `main` version.
+
+To cut a release from a commit COMMIT on `main` that declares version `A.B` (where `B` is necessarily odd), announce a new non-`main` commit that extends COMMIT merely to declare version `X.Y = next(A.B)` (where `Y` is necessarily even).
+Also immediately update the `main` version to `X.(Y+1)`.
+
+Cons:
+
+- Commits on `main` that only include patches since the previous release would still spuriously include the `Y+1` increment in the `B` dimension.
+
+## Proposal NonZero
+
+Each `main` version is degenerate; eg it is always version `0`.
+Each release version is the usual `A.B.C`.
+
+PRs do not alter the `main` version.
+
+To cut a release from a commit COMMIT on `main` (that necessarily declares version `0`), announce a new non-`main` commit that extends COMMIT merely to declare the version to be `next(A.B).0`, or `A.B.(C+1)` depending on what has changed on `main` since the previous release.
+
+Note that the degenerate versions could carry information.
+EG they could be just a single number.
+The only requirement is that they are inherently distinguished from release versions.
+
+Cons:
+
+- `main` versions would always be less than some older released version, which could cause confusion (among people and/or tools).
+
+- Cutting a release requires assessing all the changes on `main` since the last release.
+
+## Proposal Dimension
+
+FYI
+
+```
+Prelude Data.Version> makeVersion [1,2] < makeVersion [1,2,0]
+True
+```
+
+Each `main` version has only two dimensions: `A.B`.
+Each release version has at least three `A.B.C`, where `C` can be `0`.
+
+PRs do not alter the `main` version.
+
+To cut a release from a commit COMMIT on `main` that declares version `A.B`, announce a new non-`main` commit that extends COMMIT merely to declare version `A.B.0`.
+Also immediately update the `main` version to `next(A.B)`.
+
+Pros:
+
+- It enforces all desired invariants.
+
+- It's a very mechanical state machine, easy to execute and also easy to immediately recognize which state it's in.
+
+Cons:
+
+- The multi-sorted state transition system prevents any explanation from being comparatively small.
+
+- This scheme is certainly not already well-established!
+
+- It incurs spurious increments of `A.B` when the only differences between two releases were patch PRs.
+
+## Proposal Dimension124
+
+Proposal Dimension above has the downside that it incurs spurious increments of `A.B` when the only differences between two releases were patch PRs.
+The following enrichment adds the minimal amount of additional complexity to avoid that without losing any invariants.
+
+Each `main` version is `A`, `A.B` where `B>0`, or `A.B.C.2718`; it is never three-dimensional.
+Each release version is the usual `A.B.C`.
+`2718` is just a recognizable magic number; it's so large it's unlikely to come up in actual versioning and it's the first digits of _e_, which is related to _growth_, which is what the `main` branch is for.
```

Comment:
Why not just use 0?

Reply:
Yeah, this has lead to the same questions from many people.

My thought process: people will see `A.B.C.0` in a version and assume that that component can vary, since it looks so familiar. EG They'd expect `A.B.C.1`, `A.B.C.2`, etc.

I figured with `.2718` it would at least cause people to pause and ask "Wait, what? That's odd. What am I missing". (And it has, even already lol.)

Reply:
If they assume it can vary, and it won't, I think nothing goes wrong, and they can just continue on. Whereas if `.2718` makes them go "wait what?" then you're going to get _questions_.